apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,243 @@
+---
+
+<base target="_blank" />
+
+`functorch` (now `torch.func`) is a PyTorch module to provide composable transforms like JAX. It provides transforms like `grad`, `vmap`, `vjp`, etc. to allow user to easily compute Jacobian vector products and vectorize the function. And being composable, one can compute per-sample gradients simply by using `vmap(grad(model))`.
+
+Below are some of the things that I got to work on:
+
+## Adding Batching Rule for `vmap`
+`vmap` is a transform which takes a function `func` that runs on a single datapoint and returns a function which can handle a batch of data effectively vectorizing it. Semantically, it runs a `for`-loop over all data points and stacks all the results together (but does it more optimally than the for-loop version.)
+
+Example:
+```python
+import torch
+
+# Written to handle only single sample.
+# Calling it with batched input, will
+# result in incorrect output.
+def my_simple_model(input, weight, bias):
+    return input.sum(0) @ weight + bias
+
+batched_inputs = torch.randn(3, 3, 3, 3)
+weight = torch.randn(3, 1) * 5
+bias = torch.randn([])
+
+# For Loop version
+expected = []
+for input in batched_inputs:
+    expected.append(my_simple_model(input, weight, bias))
+expected = torch.stack(expected)
+
+# Vmap
+actual = torch.vmap(my_simple_model, in_dims=(0, None, None))(batched_inputs, weight, bias)
+
+# Incorrectly calling the function (silently incorrect result)
+incorrect = my_simple_model(batched_inputs, weight, bias)
+
+# Results don't match.
+assert torch.abs(expected - incorrect).sum() > 1e-3
+
+# Verify that the results match.
+torch.testing.assert_close(expected, actual)
+
+```
+
+To support `vmap` for PyTorch operators, we need to specify the batching rule i.e. how to apply the given operator for a batched input. This is similar to how PyTorch internally specifies the rule for gradient computation for operators. Batching rule is essentially a function which takes one or multiple batched inputs and computes the batched operation. In the above example to support `vmap` for `my_simple_model`, we need to know the batching rule for `torch.sum`, `@`/`torch.matmul` and `+`/`torch.add` to be able to vectorize our model. PyTorch has a lot of operators and we need to have coverage for all the operators to seamlessly support optimized `vmap` (yes, there is a `for`-loop fallback in case an operator is not supported so as not to crash the code).
+
+PyTorch operators can be very roughly categorized as primitive (internally CompositeExplicitAutograd) vs composite (internally CompositeImplicitAutograd). Composite operators are derived from the primitive operators. So, to have complete coverage, we need to have batching rules for all the primitive operators and we get the rules for composite operators for free.
```

Comment:
nit. very roughly -> roughly

Reply:
nit. Put Composite* works within backticks.

Reply:
What does it mean that a "composite operator is derived from a primitive operator"?

Reply:
Don't start a sentence with `so`. In this case, just drop the so or use another connector like "for this reason"

Reply:
The cadence of the last sentence is a bit off. You are saying "to have A, we need B and we get A". Consider rewriting it